mathics/builtin/lists.py

```diff
@@ -2012,3 +2012,113 @@ def get_call():
                                              evaluation, same_test))
         result.sort()
         return result
+
+class Fold(Builtin):
+    """
+    <dl>
+    <dt>'Fold[$expr$, $x$, $list$]'
+        <dd>Expression on all elements of list, with initial value of x.
+    <dt>'Fold[$expr$, $list$]'
+        <dd>The same as Fold[$expr$, First[list], Rest[list]]
+    </dl>
+
+    >> Fold[Plus, 5, {1, 1, 1}]
+     = 8
+    >> Fold[f, 5, {1, 2, 3}]
+     = f[f[f[5, 1], 2], 3]
+    """
+
+    attributes = ('NHoldRest',)
+
+    rules = {
+        'Fold[exp_, x_, list_List]': 'Module[{},res=x; Do[res = exp[res, list[[i]]], {i, 1, Length[list]}]; res]',
+        'Fold[exp_, list_List]': u'Fold[exp, First[list], Rest[list]]',
+    }
+
+
+class FoldList(Builtin):
+    """
+    <dl>
+    <dt>'FoldList[$expr$, $x$, $list$]'
+        <dd>
+            Apply expr successive on all elements of list, and return list, where $x$ is the first element.
+        </dd>
+    </dt>
+    <dt>FoldList[$expr$, $list$]'
+        <dd>
+            The same as FoldList[$expr$, First[list], Rest[list]].
+        </dd>
+        </dt>
+    </dl>
+    >> FoldList[f, x, {1, 2, 3}]
+     = {x, f[x, 1], f[f[x, 1], 2], f[f[f[x, 1], 2], 3]}
+    >> FoldList[Times, {1, 2, 3}]
+     = {1, 2, 6}
+    """
+
+    rules = {
+        'FoldList[exp_, x_, list_List]': 'Prepend[Table[Fold[exp, x, Take[list, i]], {i, 1, Length[list]}], x]',
+        'FoldList[exp_, list_List]': 'FoldList[exp, First[list], Rest[list]]',
+    }
+
+
+class Accumulate(Builtin):
+    """
+    <dl>
+    <dt>'Accumulate[$list$]'
+        <dd>
+        Accumulate values from list and return new list
+        </dd>
+        </dt>
+    </dl>
+    >> Accumulate[{1, 2, 3}]
+     = {1, 3, 6}
+    """
+
+    rules = {
+        'Accumulate[list_List]': 'FoldList[Plus, list]'
```

Comment:
According to the Mathematica documentation, "`Accumulate[list]` is effectively equivalent to `Rest[FoldList[Plus,0,list]]`."

I see two minor issues with this approach:

- `Accumulate` should work for any head (not just `List`). e.g.
```
In[1] := Accumulate[f[a, b, c, d]]
Out[1]= f[a, a + b, a + b + c, a + b + c + d]
```

- This fails on the empty list
```
In[3]:= Accumulate[{}]
Traceback (most recent call last):
  File "mathics/main.py", line 324, in <module>
    main()
  File "mathics/main.py", line 315, in main
    shell.evaluate(total_input)
  File "mathics/main.py", line 120, in evaluate
    out_callback=out_callback)
  File "/home/angus/prog/Mathics/mathics/core/evaluation.py", line 207, in __init__
    result = evaluate()
  File "/home/angus/prog/Mathics/mathics/core/evaluation.py", line 192, in evaluate
    result = query.evaluate(self)
  File "mathics/core/expression.py", line 776, in mathics.core.expression.Expression.evaluate (mathics/core/expression.c:22266)
    result = result.evaluate(evaluation)
  File "mathics/core/expression.py", line 776, in mathics.core.expression.Expression.evaluate (mathics/core/expression.c:22266)
    result = result.evaluate(evaluation)
  File "mathics/core/expression.py", line 726, in mathics.core.expression.Expression.evaluate (mathics/core/expression.c:21079)
    leaves[index] = leaves[index].evaluate(evaluation)
  File "mathics/core/expression.py", line 773, in mathics.core.expression.Expression.evaluate (mathics/core/expression.c:22179)
    result = rule.apply(new, evaluation, fully=False)
  File "mathics/core/rules.py", line 115, in mathics.core.rules.BaseRule.apply (mathics/core/rules.c:2677)
    self.pattern.match(
  File "mathics/core/pattern.py", line 224, in mathics.core.pattern.ExpressionPattern.match (mathics/core/pattern.c:6424)
    self.head.match(
  File "mathics/core/pattern.py", line 153, in mathics.core.pattern.AtomPattern.match (mathics/core/pattern.c:4758)
    yield_func(vars, None)
  File "mathics/core/pattern.py", line 216, in mathics.core.pattern.ExpressionPattern.match.yield_head (mathics/core/pattern.c:6024)
    self.get_pre_choices(
  File "mathics/core/pattern.py", line 341, in mathics.core.pattern.ExpressionPattern.get_pre_choices (mathics/core/pattern.c:8974)
    yield_func(vars)
  File "mathics/core/pattern.py", line 201, in mathics.core.pattern.ExpressionPattern.match.yield_choice (mathics/core/pattern.c:5833)
    self.match_leaf(
  File "mathics/core/pattern.py", line 506, in mathics.core.pattern.ExpressionPattern.match_leaf (mathics/core/pattern.c:12726)
    self.get_wrappings(
  File "mathics/core/pattern.py", line 362, in mathics.core.pattern.ExpressionPattern.get_wrappings (mathics/core/pattern.c:9936)
    yield_func(items[0])
  File "mathics/core/pattern.py", line 502, in mathics.core.pattern.ExpressionPattern.match_leaf.yield_wrapping (mathics/core/pattern.c:11242)
    leaf.match(match_yield, item, vars, evaluation, fully=True,
  File "mathics/builtin/patterns.py", line 539, in mathics.builtin.patterns.Pattern_.match (mathics/builtin/patterns.c:9456)
    self.pattern.match(yield_func, expression, new_vars, evaluation)
  File "mathics/builtin/patterns.py", line 701, in mathics.builtin.patterns.Blank.match (mathics/builtin/patterns.c:12459)
    yield_func(vars, None)
  File "mathics/core/pattern.py", line 499, in mathics.core.pattern.ExpressionPattern.match_leaf.match_yield (mathics/core/pattern.c:11104)
    yield_func(new_vars, items_rest)
  File "mathics/core/rules.py", line 74, in mathics.core.rules.BaseRule.apply.yield_match (mathics/core/rules.c:2129)
    new_expression = self.do_replace(vars, options, evaluation)
  File "mathics/core/rules.py", line 165, in mathics.core.rules.BuiltinRule.do_replace (mathics/core/rules.c:4029)
    return self.function(evaluation=evaluation, **vars_noctx)
  File "/home/angus/prog/Mathics/mathics/builtin/lists.py", line 777, in apply
    return expr.leaves[0]
IndexError: list index out of range
```